fix: Overflow-detection correctness issues in double-delta and bit-width-reduction filters

rroelke opened the pull request:
[SC-65154](https://app.shortcut.com/tiledb-inc/story/65154/overflow-buffer-on-read) surfaced an issue in a filter pipeline where data would be written, but not read back properly.  Investigating identified that this could happen in the event of arithmetic overflow.  Both the double delta and bit width reduction filters have errors in their overflow detection.

The issue reporter's filter pipeline fed double delta into bit width reduction.  Since double delta produces essentially binary blobs as output containing sign bits, this unsurprisingly produces very wide bit widths.  And apparently arithmetic overflows.  This caused the bit width reduction to write down some incorrect bits, which un-filtered into invalid input to the double delta filter.

The fix of course is to correct the overflow detection.  The C++ standard library does not provide routines to do this, so we must write our own and assert their correctness.

~~That brings in the more significant part of this pull request, which is the introduction of Rust into core library builds.  The Rust standard library *does* provide overflow detection routines.  We can write our own in C++ (which can be in-lined), and then write unit tests which compare their output against code written in Rust.~~

~~As such this is a light step towards using Rust inside of the core library - this pull request only introduces small functions in test code, which we hope will enable us to divorce the build configuration and CI details from the novelty of doing something interesting in Rust.~~

EDIT: the Rust bits are all reverted so we can merge the bug fix (which is awaited upstream and will be a good candidate for 2.28) from the Rust component (which is not desired for 2.28)

---
TYPE: BUG | IMPROVEMENT
DESC: Fix overflow detection errors in double delta and bit width reduction filters

rroelke commented on test/support/oxidize/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+set(CARGO_LIB_DIR "${CMAKE_CURRENT_BINARY_DIR}/target/${CARGO_TARGET}")
+
+if (TILEDB_RUST)
+    foreach(crate IN LISTS TILEDB_CRATES)
```
My main reason was that if I build a particular unit test target (which is usually what I find myself doing) then I only want to build the crates which that particular target needs.

The single crate doing `--all-targets` is definitely not what I want for that

davisp replied:
Also, it occurs to me that as long as everything shares a single workspace we should still benefit from not recompiling dependencies. I bet we could turn this into a function call as well and then our crate list would look something like:

```cmake
add_crate(
    NAME
        arithmetic
    SOURCES
        src/lib.rs
        src/utils.rs
)
```

rroelke replied:
46503afdf9122d688821b40a4ca79e0c4c76eb6a

rroelke commented on test/support/oxidize/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+                cp "${CARGO_INCLUDE_DIR}/cxxbridge/${crate}/src/lib.rs.h" "${CMAKE_CURRENT_BINARY_DIR}/${crate}.h"
+            USES_TERMINAL
+            DEPENDS
+                ${CMAKE_CURRENT_SOURCE_DIR}/${crate}/src/lib.rs
```
I did foresee that problem but decided it was a problem for later :)

I imagined that we could use a glob here.  From what I'm reading now, the `cmake` documentation doesn't really recommend this because it doesn't pick up files that are added, but there is an [easy workaround to that](https://www.reddit.com/r/cmake/comments/17j3fmu/comment/k717tlm/).

If we can list files per crate I'm fine with that, but I would like to avoid a single monolithic list that would trigger a re-build of all of the crates

davisp replied:
I attempted the glob approach before switching back to the list of files. I didn't look too hard into it, but using the glob list appeared to unconditionally rerun Cargo on every build rather than only when source files changed.

rroelke replied:
46503afdf9122d688821b40a4ca79e0c4c76eb6a implements the `add_crate` function with a manual list of source files. I think that should be sufficient to resolve this

rroelke commented on .github/workflows/ci-linux_mac.yml:
```diff
@@ -46,6 +46,11 @@ on:
+        default: false
+        description: 'Enable Rust'
+        required: false
+        type: boolean
```
I don't think I added it for anything yet other than the standalone unit tests, since it wouldn't do anything yet except for those

rroelke commented on test/support/CMakeLists.txt:
```diff
@@ -115,4 +115,6 @@ commence(object_library mem_helpers)
   this_target_object_libraries(baseline)
 conclude(object_library)
 
+add_subdirectory(oxidize)
```
There's quite a bit of unit test code under `tiledb` that references test code living in `test` so there is some back and forth.

I was picturing two workspaces, `test/support/oxidized` in addition to `tiledb/oxidized`

I suppose we could have one workspace with `tiledb/oxidized/tiledb` and `tiledb/oxidized/test` sub-directories

davisp replied:
My main priority is just that all Rust code exists in a single workspace for the dependency building reason. Beyond that I don't mind too much other than I'd find it rather odd to be shipping things that are built in the test directory even though there's no real reason that'd be an issue.

rroelke replied:
11e4103ac616e10a5dcba009107c040855d614f2 moves to repository root `oxidize` directory

rroelke commented on test/support/oxidize/arithmetic/src/lib.rs:
```diff
@@ -0,0 +1,89 @@
+
+#[cxx::bridge]
+mod ffi {
+    #[namespace = "rust::std"]
```
Oops, this was also transitional, I intended to rename this to `rust::arithmetic`

davisp replied:
I'd prefer to have `tiledb` in the namespace somewhere to indicate that its code we're exposing. Though feel free to do a `using rust = tiledb::rust` or similar.

rroelke replied:
Changed to `tiledb::rust::arithmetic` in 7e9cd08462d0734ac0d67eb5edae0238758b7360

rroelke commented on tiledb/sm/compressors/dd_compressor.cc:
```diff
@@ -64,7 +65,7 @@ void DoubleDelta::compress(
     case Datatype::GEOM_WKB:
     case Datatype::GEOM_WKT:
-      return DoubleDelta::compress<std::byte>(input_buffer, output_buffer);
+      return DoubleDelta::compress<uint8_t>(input_buffer, output_buffer);
```
Since the first thing that happens to the data is cast to `int64_t`, we don't.  The reason I had to change from `std::byte` is because of some of the template code I added in `tiledb/common/arithmetic.h`.  It doesn't feel right to add a specialization for `std::byte` to those since from what I read `std::byte` is specifically not intended for arithmetic.

rroelke commented on test/support/oxidize/arithmetic/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+version.workspace = true
+
+[dependencies]
+cxx = { workspace = true }
```
Hmm, I actually prefer the `{ workspace = true }` style. I am sure we both agree that they should be consistent however!

rroelke replied:
c18a374a5f5051deaea66c5549fe824531c85e92

davisp commented on test/support/oxidize/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+                "--manifest-path" ${CARGO_MANIFEST_PATH}
+                "-p" ${crate}
+            COMMAND
+                cp "${CARGO_INCLUDE_DIR}/cxxbridge/${crate}/src/lib.rs.h" "${CMAKE_CURRENT_BINARY_DIR}/${crate}.h"
```
Ah, I didn't see the rename part until you pointed it out which is fair as the `#include "path/to/lib.rs.h"` is a bit awkward.

However, for cross platform support, you'll probably want to remove this `COMMAND` and replace it with the [`file` function](https://cmake.org/cmake/help/latest/command/file.html).

rroelke replied:
fa19829d71de25fc8e649d7abfd027f84909e53d

rroelke commented on tiledb/common/test/unit_arithmetic.cc:
```diff
@@ -0,0 +1,304 @@
+  };
+
+  SECTION("Rapidcheck") {
+    rc::prop(
+        "checked_arithmetic<uint32_t>::add", [doit](uint32_t a, uint32_t b) {
+          doit.operator()<tiledb::test::AsserterRapidcheck>(a, b);
+        });
```
My assumption is that function calls across the FFI boundary will not be inlined where as the C++ template ones would be

rroelke commented on tiledb/common/test/unit_arithmetic.cc:
```diff
@@ -0,0 +1,304 @@
+    std::optional<int32_t> rs;
+    {
+      int32_t value;
+      if (rust::std::i32_checked_add(a, b, value)) {
```
[This](https://github.com/dtolnay/cxx/issues/87#issuecomment-2677287152) was a pretty interesting comment about this

rroelke commented on tiledb/sm/compressors/dd_compressor.h:
```diff
@@ -134,10 +136,10 @@ class DoubleDelta {
    */
   template <class T>
-  static void compute_bitsize(T* in, uint64_t num, unsigned int* bitsize);
+  static unsigned compute_bitsize(const T* in, uint64_t num);
```
503d2dd3a1278faa6050fe352294fed0299f78cb

rroelke commented on tiledb/sm/compressors/dd_compressor.cc:
```diff
@@ -257,35 +257,54 @@ void DoubleDelta::compress(ConstBuffer* input_buffer, Buffer* output_buffer) {
   }
+
   // Calculate bitsize of the maximum absolute double delta
+  unsigned bitsize = 0;
```
a0690b53a825b0dd230937c4af8f5ff7fee3c3db

rroelke commented on .github/workflows/check-formatting.yml:
```diff
@@ -38,3 +38,25 @@ jobs:
+    timeout-minutes: 90
+    name: Check Rust Format
+    steps:
+      - name: Install Rust Stable
```
20280b42ad5223145ecc86a8f0f2484f469e9425 hopefully fixes this...

rroelke commented on test/support/oxidize/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+)
+
+if (NOT TILEDB_TESTS)
+  message(FATAL_ERROR "Building Rust without tests is not supported.")
```
Looking at your original branch it was the same way there.  But I'm not sure we need this at all?  If there's no tests then I expect there just aren't any dependencies on these targets.

I removed this in 5fae3b8f75fa07f5860f80dcd16dce538c9bfb00 but we can add something back if we do need it

rroelke commented on test/support/oxidize/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+members = [
+  "arithmetic"
+]
+default-members = [
```
d186de573142633fcf265b2a5a7c31d8afb7b050

rroelke commented:
@davisp I removed the Rust components here, after discussing with @ihnorton we agreed that we want the filter bug fix to land in 2.28 if possible but without introducing any Rust build onto that branch.

So I've updated this to just be the overflow detection, filter fixes, and unit tests for those.

Rust will follow once this is merged

teo-tsirpanis commented on tiledb/sm/compressors/dd_compressor.h:
```diff
@@ -35,6 +35,8 @@
 
 #include "tiledb/common/common.h"
 
+#include <type_traits>
+
```
Why is this needed?

rroelke replied:
Vestigial from an intermediate state. Removed e212f5

teo-tsirpanis commented on tiledb/common/arithmetic.h:
```diff
@@ -0,0 +1,251 @@
+struct extended_integral_type<uint32_t> {
+  using type = uint64_t;
+};
+
```
```suggestion

template <class T> using extended_integral_type_t = typename extended_integral_type<T>::type;

```

Using `extended_integral_type_t` will slightly shorten some code below.

rroelke replied:
756c11d

teo-tsirpanis commented on tiledb/sm/compressors/dd_compressor.cc:
```diff
@@ -257,35 +257,54 @@ void DoubleDelta::compress(ConstBuffer* input_buffer, Buffer* output_buffer) {
+    const DeltaType cur = (DeltaType)(in[i]);
+    const DeltaType prev = (DeltaType)(in[i - 1]);
+
+    const std::optional<int64_t> cur_delta =
+        common::checked_arithmetic<DeltaType>::sub_signed(cur, prev);
+    if (!cur_delta.has_value()) {
+      throw DoubleDeltaException(
+          "Cannot compress with DoubleDelta: delta exceeds range of int64_t");
+    }
+
+    if (prev_delta.has_value()) {
+      const std::optional<int64_t> dd =
+          common::checked_arithmetic<int64_t>::sub(
+              cur_delta.value(), prev_delta.value());
+      if (!dd.has_value()) {
+        throw DoubleDeltaException(
+            "Cannot compress with DoubleDelta; Some negative double delta is "
+            "out "
+            "of bounds");
+      }
```
Can't we wrap around the result? It is weird for a filter to be unable to process some inputs.

rroelke replied:
There is precedent here in the previous code.
```
  // Handle error
  if (delta_out_of_bounds) {
    throw DoubleDeltaException(
        "Cannot compress with DoubleDelta; Some negative double delta is out "
        "of bounds");
  }
```
i.e. it's not a new notion that this filter cannot process all input data.

teo-tsirpanis replied:
~~Can we fix it now?~~ Maybe not, it might require a format version bump. 😩

teo-tsirpanis replied:
But since we already change the logic (especially in the bit width reduction filter), a format version bump might be warranted either way, and we should seize the opportunity to get it right.

rroelke replied:
RE: format version bump

I think we don't actually need this since we changed only the forward filter logic. Since un-filtering is not changed, any bytes that were written by prior versions of this filter will be un-filtered in the same way.

RE: passing data which hits the overflow through the filter

We should not bother implementing this.
1) this filter is designed to be used with offsets of similar-size strings, where it gets the best encoding results.  It is meant to be used at the beginning of the filter pipeline on non-descending non-negative integers.  It doesn't really matter how it behaves on other streams because that's not where it provides business value.
2) if there is an arithmetic overflow then the double delta produces a value with a high bit width which it will then use to encode each of the deltas in the stream.  This will result in data which is not compacted relative to the input (and is possibly larger) but also requires additional CPU overhead to decode.  Hence streams which have this overflow in them are better off not applying this encoding at all.
3) You could perhaps work around (2) by using windows (as the bit width reduction filter does).  But this will produce worse results against the target use case of (1), whether it would be a worse encoding due to window padding, or a performance hit to identify a dynamic window size, or both.

teo-tsirpanis replied:
> It doesn't really matter how it behaves on other streams because that's not where it provides business value.

I don't believe filters have the "right" to reject input, even if they are documented to work on special forms of data. Not hard-failing when writing data does arguably have business value, and I believe somebody will inevitably hit this error one day, and fixing it then would be more urgent than fixing it now.

> This will result in data which is not compacted relative to the input (and is possibly larger) but also requires additional CPU overhead to decode.

This is not an unusual thing with filters; if you compress high-entropy data with zstd, you will not get size savings but the operation will still succeed, which is the improtant thing IMO.

> You could perhaps work around (2) by using windows (as the bit width reduction filter does).

We already split data before filtering them, both at the tile and the chunk level (I'm not very happy that the BWR filter does it again but that's another story) so the impact of filtering of suboptimal data is not globalized over the whole fragment or tile.

rroelke replied:
> Not hard-failing when writing data does arguably have business value, and I believe somebody will inevitably hit this error one day, and fixing it then would be more urgent than fixing it now

If someone hits this error when using the double delta filter then that's a good sign that they should not be using the double delta filter, and if this does show up in the field then that's the guidance we should provide.

(I'll note that for the story in question the double delta was actually fine, the problem was the overflow detection in the bit width reduction filter applied after the double delta filter... another example of something that shouldn't really be done)

> This is not an unusual thing with filters

It certainly *can* happen, but it's not desirable, and what we're discussing here amounts to a request to enable it.


With respect to this PR it's all moot.  The unresolved question here is about the format version.

> I think we don't actually need this since we changed only the forward filter logic. Since un-filtering is not changed, any bytes that were written by prior versions of this filter will be un-filtered in the same way.

Do you agree?

teo-tsirpanis replied:
I agree.

teo-tsirpanis reviewed (approved):
Thanks!

davisp commented on tiledb/sm/compressors/dd_compressor.cc:
```diff
@@ -257,35 +257,54 @@ void DoubleDelta::compress(ConstBuffer* input_buffer, Buffer* output_buffer) {
+      if (!dd.has_value()) {
+        throw DoubleDeltaException(
+            "Cannot compress with DoubleDelta; Some negative double delta is "
+            "out "
+            "of bounds");
```
Seems like an odd line break here.

davisp reviewed (approved):
+1. I've managed to convince myself I understand how all of the math works which was not a short process. Not gonna lie, I was a lot more confident when there was a Rust version to check against (I completely understand why that got removed though). Still, this all looks good to me.

rroelke commented:
> Not gonna lie, I was a lot more confident when there was a Rust version to check against 

When I have it in me to deal with CI again I'll put up a new PR which adds the Rust tests back
